## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [419]:
import pandas as pd
pd.set_option('display.float_format', lambda x: '%.2f' % x)
src_data_path = r'Resources/client_dataset.csv'
df = pd.read_csv(src_data_path)
df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year         item_id     category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   
6   8080280  2023-04-19          16        2023  UAC50632-25-4E  consumables   
7   8946782  2023-02-14           7        2023  UEU25316-01-1U  consumables   
8   9394118  2023-05-13          19        2023  DDA06248-98-7Z  consumables   
9   9665852  2023-04-14          15        2023  DUU85042-87-5E    furniture   

          subcategory  unit_price  unit_cost  unit_weight  qty  line_number  
0            wall art     1096.80     762.71         7.50  105            1  
1                pens       24.95      15.09         1.49   21            0  
2  project management       13.52       7.86         1.68   39            6  
3                pens       36.42      24.85         1.23   29            3  
4                misc      195.10     108.17        46.43   20            1  
5              plants      172.84     154.44         2.79   78            0  
6    kitchen supplies      245.56     231.95         1.73   56            5  
7                pens       20.99      16.97         5.12   22            4  
8      paper products       83.86      59.00         0.27   49            3  
9             seating       66.88      54.15         2.47   78            1

In [420]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [421]:
# Use the describe function to gather some basic statistics
df.describe


<bound method NDFrame.describe of            first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id

#### Use this space to do any additional research and familiarize yourself with the data.

- Overall Info About the Client and their Purchase habits

In [422]:
client_info = df[['client_id','first', 'last', 'phone', 'email', 
                          'job', 'order_id', 'order_date', 'order_week',  'item_id', 'category', 'subcategory', 'unit_price', 
                          'unit_cost', 'unit_weight', 'qty', 'line_number']]
client_info

client_id      first       last               phone  \
0          58515     Donald    Harding  793-904-7725x39308   
1          37609    Tiffany      Myers    201.442.4543x942   
2          57113    Shannon     Watson   687.737.9424x8503   
3          46554     Nathan      Baker    827-788-8123x012   
4          92089  Christina   Schwartz        265-829-3643   
...          ...        ...        ...                 ...   
54634      29043     Lauren      Reese  977.877.5272x11382   
54635      41908    Derrick      Moore        358.661.5483   
54636      35176     Monica  Gutierrez    294.805.9100x339   
54637      24485      Wanda    Solomon       (311)767-4924   
54638      24007     Dwayne       Bell   606-619-7667x9508   

                                     email                      job  order_id  \
0         harding.donald.7185@sullivan.com             Immunologist   8953482   
1        myers.t.6537@ferguson-johnson.net          Music therapist   8069089   
2                    swatson8146@payne.net             Immunologist   1902144   
3                   bakernathan@benson.com    Accounting technician   9031802   
4      christinaschwartz9252@mcconnell.com             Chiropractor   1322274   
...                                    ...                      ...       ...   
54634   late_reese_4081@montoya-chavez.org           Radio producer   9021716   
54635         derrick.moore.2602@pope.info                 Musician   6290153   
54636            gutierrezm3195@morris.org         Graphic designer   8692622   
54637            solomonwanda5962@ross.org             Toxicologist   7592730   
54638                     belld@galvan.com  Chief Financial Officer   7489403   

       order_date  order_week         item_id     category  \
0      2023-04-28          17  EUD29711-63-6U        decor   
1      2023-05-19          20  XDA18116-89-4A  consumables   
2      2023-01-29           4  ABE59463-05-7E     software   
3      2023-04-25          17  ZMM00836-65-0C  consumables   
4      2023-05-28          21  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634  2023-04-26          17  UCA28532-31-0B     software   
54635  2023-02-08           6  EBD29277-23-7U  consumables   
54636  2023-03-05           9  EBD30093-55-9Z  consumables   
54637  2023-01-18           3  XMZ38239-49-3D    furniture   
54638  2023-01-04           1  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  
0                1  
1                0  
2                6  
3                3  
4                1  
...            ...  
54634            8  
54635            0  
54636            6  
54637            0  
54638            5  

[54639 rows x 17 columns]

- Client Personal Identifiable Info

In [423]:

client_pii = df[['client_id','first', 'last', 'phone', 'email', 'job']]
client_pii


client_id      first       last               phone  \
0          58515     Donald    Harding  793-904-7725x39308   
1          37609    Tiffany      Myers    201.442.4543x942   
2          57113    Shannon     Watson   687.737.9424x8503   
3          46554     Nathan      Baker    827-788-8123x012   
4          92089  Christina   Schwartz        265-829-3643   
...          ...        ...        ...                 ...   
54634      29043     Lauren      Reese  977.877.5272x11382   
54635      41908    Derrick      Moore        358.661.5483   
54636      35176     Monica  Gutierrez    294.805.9100x339   
54637      24485      Wanda    Solomon       (311)767-4924   
54638      24007     Dwayne       Bell   606-619-7667x9508   

                                     email                      job  
0         harding.donald.7185@sullivan.com             Immunologist  
1        myers.t.6537@ferguson-johnson.net          Music therapist  
2                    swatson8146@payne.net             Immunologist  
3                   bakernathan@benson.com    Accounting technician  
4      christinaschwartz9252@mcconnell.com             Chiropractor  
...                                    ...                      ...  
54634   late_reese_4081@montoya-chavez.org           Radio producer  
54635         derrick.moore.2602@pope.info                 Musician  
54636            gutierrezm3195@morris.org         Graphic designer  
54637            solomonwanda5962@ross.org             Toxicologist  
54638                     belld@galvan.com  Chief Financial Officer  

[54639 rows x 6 columns]

- Client Order Info

In [424]:

order_info = df[['order_id', 'order_date', 'order_week','unit_price', 
                          'unit_cost', 'unit_weight', 'qty']]
order_info

order_id  order_date  order_week  unit_price  unit_cost  unit_weight  \
0       8953482  2023-04-28          17     1096.80     762.71         7.50   
1       8069089  2023-05-19          20       24.95      15.09         1.49   
2       1902144  2023-01-29           4       13.52       7.86         1.68   
3       9031802  2023-04-25          17       36.42      24.85         1.23   
4       1322274  2023-05-28          21      195.10     108.17        46.43   
...         ...         ...         ...         ...        ...          ...   
54634   9021716  2023-04-26          17       83.13      51.60         2.25   
54635   6290153  2023-02-08           6      206.59     175.46        11.70   
54636   8692622  2023-03-05           9       65.66      57.31         4.16   
54637   7592730  2023-01-18           3        1.48       1.22        18.04   
54638   7489403  2023-01-04           1        3.01       1.61         2.07   

        qty  
0       105  
1        21  
2        39  
3        29  
4        20  
...     ...  
54634    33  
54635    47  
54636   475  
54637   112  
54638  1031  

[54639 rows x 7 columns]

- Product Information

In [425]:
item_info = df[['line_number', 'item_id', 'category', 'subcategory']]
item_info




line_number         item_id     category         subcategory
0                1  EUD29711-63-6U        decor            wall art
1                0  XDA18116-89-4A  consumables                pens
2                6  ABE59463-05-7E     software  project management
3                3  ZMM00836-65-0C  consumables                pens
4                1  BZX55559-12-3X  consumables                misc
...            ...             ...          ...                 ...
54634            8  UCA28532-31-0B     software           logistics
54635            0  EBD29277-23-7U  consumables    kitchen supplies
54636            6  EBD30093-55-9Z  consumables   bathroom supplies
54637            0  XMZ38239-49-3D    furniture              stands
54638            5  ZCZ05589-50-5D     software  project management

[54639 rows x 4 columns]

In [426]:
# What three item categories had the most entries?
most_cat_entries = item_info['category'].value_counts().head(3)
most_cat_entries

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [427]:
# For the category with the most entries, which subcategory had the most entries?

# most_sub_cat_entries = pd.concat(most_cat_entries, item_info['subcategory'])
most_subcat_entries = item_info['subcategory'].value_counts().head(1)
most_subcat_entries

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [428]:
# Which five clients had the most entries in the data?
most_five_client_entries = client_pii['client_id'].value_counts().head(5)
most_five_client_entries

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [429]:
# Store the client ids of those top 5 clients in a list.
list_top_fiveclients = list(most_five_client_entries.index)
list_top_fiveclients

[33615, 66037, 46820, 38378, 24741]

In [430]:
# How many total units (the qty column) did the client with the most entries order order?
clientID_Max = client_pii['client_id'].value_counts().idxmax()
clientID_Max
clientID = df[client_pii['client_id'] == clientID_Max]
total_qty = clientID['qty'].sum()
print(f'Client ID {clientID_Max} ordered the most items which sums up to {total_qty} quantities')


Client ID 33615 ordered the most items which sums up to 64313 quantities


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [431]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df1 = df.copy()
df1['subtotal'] = (order_info['unit_price'] * order_info['qty'])
df1

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  subtotal  
0                1 115164.00  
1                0    523.95  
2                6    527.28  
3                3   1056.18  
4                1   3902.00  
...            ...       ...  
54634            8   2743.29  
54635            0   9709.73  
54636            6  31188.50  
54637            0    165.76  
54638            5   3103.31  

[54639 rows x 19 columns]

In [432]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def assign_shipping_prices(item_weight):
    match item_weight:
        case num if num > 50:
            item_weight = item_weight * 7
        case num if num <= 50:
            item_weight = item_weight * 10
    return item_weight
df1['shipping_price'] = order_info['unit_weight'].apply(assign_shipping_prices)
df1


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  subtotal  shipping_price  
0                1 115164.00           75.00  
1                0    523.95           14.90  
2                6    527.28           16.80  
3                3   1056.18           12.30  
4                1   3902.00          464.30  
...            ...       ...             ...  
54634            8   2743.29           22.50  
54635            0   9709.73          117.00  
54636            6  31188.50           41.60  
54637            0    165.76          180.40  
54638            5   3103.31      

In [433]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
tax_rate = 9.25 / 100
print(tax_rate)
df1['total_price'] = (df1['subtotal'] + df1['shipping_price']) * (1 + tax_rate)
df1


0.0925


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...     category         subcategory  \
0              17        2023  ...        decor            wall art   
1              20        2023  ...  consumables                pens   
2               4        2023  ...     software  project management   
3              17        2023  ...  consumables                pens   
4              21        2023  ...  consumables                misc   
...           ...         ...  ...          ...                 ...   
54634          17        2023  ...     software           logistics   
54635           6        2023  ...  consumables    kitchen supplies   
54636           9        2023  ...  consumables   bathroom supplies   
54637           3        2023  ...    furniture              stands   
54638           1        2023  ...     software  project management   

      unit_price  unit_cost  unit_weight   qty  line_number  subtotal  \
0        1096.80     762.71         7.50   105            1 115164.00   
1          24.95      15.09         1.49    21            0    523.95   
2          13.52       7.86         1.68    39            6    527.28   
3          36.42      24.85         1.23    29            3   1056.18   
4         195.10     108.17        46.43    20            1   3902.00   
...          ...        ...          ...   ...          ...       ...   
54634      83.13      51.60         2.25    33            8   2743.29   
54635     206.59     175.46        11.70    47            0   9709.73   
54636      65.66      57.31         4.16   475            6  31188.50   
54637       1.48       1.22        18.04   112            0    165.76   
54638       3.01       1.61         2.07  1031            5   3103.31   

       shipping_price  total_price  
0               75.00    125898.61  
1               14.90       588.69  
2               16.80       594.41  
3               12.30      1167.31  
4              464.30      4770.18  
...               ...          ...  
54634           22.50      3021.63  
54635          117.00     10735.70  
54636           41.60     34118.88  
54637          180.40       378.18  
54638       

In [434]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df1['item_cost'] = (order_info['unit_cost'] * order_info['qty']) + df1['shipping_price']
df1


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number  subtotal  shipping_price  total_price  \
0             7.50   105            1 115164.00           75.00    125898.61   
1             1.49    21            0    523.95           14.90       588.69   
2             1.68    39            6    527.28           16.80       594.41   
3             1.23    29            3   1056.18           12.30      1167.31   
4            46.43    20            1   3902.00          464.30      4770.18   
...            ...   ...          ...       ...             ...          ...   
54634         2.25    33            8   2743.29           22.50      3021.63   
54635        11.70    47            0   9709.73          117.00     10735.70   
54636         4.16   475            6  31188.50           41.60     34118.88   
54637        18.04   112            0    165.76          180.40       378.18   
54638         2.07  1031            5   3103.31           20.70      3412.98   

       item_cost  
0       80159.55  
1         331.79  
2         323.34  
3         732.95  
4        2627.70  
...          ...  
54634    1725.30  
54635    8363.62  
54636   27263.85  
54637     317.04  
54638    1680.61  

[54639 row

In [435]:
# Create a column for the profit of each line using line cost and line price
df1['profit'] = df1['total_price'] - df1['item_cost']
df1['profit'] = df1['profit']


df1

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number  subtotal  shipping_price  total_price  item_cost   profit  
0                1 115164.00           75.00    125898.61   80159.55 45739.06  
1                0    523.95           14.90       588.69     331.79   256.90  
2                6    527.28           16.80       594.41     323.34   271.07  
3                3   1056.18           12.30      1167.31     732.95   434.36  
4                1   3902.00          464.30      4770.18    2627.70  2142.48  
...            ...       ...             ...          ...        ...      ...  
54634            8   2743.29           22.50      3021.63    1725.30  1296.33  
54635            0   9709.73          117.00     10735.70    8363.62  2372.08  
54636            6  31188.50           41.60     34118.88   27263.85  6855.03  
54637            0    165.76          180.40       378.18     317.04    61.14  
54638            5   3103.31           20.70      3412.98    1680.61  1732.37  

[54639 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [436]:
# Check your work using the totals above
target_order = {
    'order_id': [2742071, 2173913, 6128929],
    'reciept_total_price': [15281189, 16238871, 92344125]

}
consolidate_orders = df1.groupby('order_id')['total_price'].sum().reset_index()
consolidate_orders.rename(columns={'total_price' : 'Total Price'}, inplace=True)
target_order_data = pd.DataFrame(target_order)
comparison = pd.merge(consolidate_orders, target_order_data, on='order_id')
comparison['Match'] = abs(comparison['Total Price'] - comparison['reciept_total_price']) < 1e-2
comparison



order_id  Total Price  reciept_total_price  Match
0   2173913    119838.36             16238871  False
1   2742071    139844.51             15281189  False
2   6128929    752748.75             92344125  False

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [437]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
amt_spent = []
for client in list_top_fiveclients:
    client_data = df1.loc[df1['client_id'] == client]
    total_spent_qty = client_data['qty'].sum()
    total_spent = client_data['total_price'].sum()
    amt_spent.append({'client_id': client, 'qty_spent': total_spent_qty, 'total_spent': round(total_spent, 2)})
total_qty_spent = pd.DataFrame(amt_spent)
print(total_qty_spent)

   client_id  qty_spent  total_spent
0      33615      64313   6392918.35
1      66037      43018   8746255.31
2      46820      75768   8006332.53
3      38378      73667   9172866.14
4      24741     239862  76678479.05


In [438]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.
pd.set_option('display.float_format', lambda x: '%.2f' % x)
client_summary = df1.groupby('client_id').agg({
    'qty' : 'sum',
    'shipping_price': 'sum',
    'subtotal': 'sum',
    'profit': 'sum'
}).reset_index()
client_summary = client_summary.round(2)
top_client_summary = client_summary.sort_values(by='subtotal', ascending=False).head(5)
top_client_summary

client_id      qty  shipping_price     subtotal       profit
758      78965  4033302         7383.30 595047041.80 246859505.41
840      85931   644423         2763.00 207192783.92 108358048.60
320      38473   402770         7832.20 163620503.65  42946283.31
850      86721  1010426         6154.60 101583079.67  37024209.98
821      84344   466968         6944.00  92871825.66  27273283.52

In [439]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
df1

# Define a function that converts a dollar amount to millions.
def currency_format_millions(amt):
    return amt/1_000_000

# # Apply the currency_format_millions function to only the money columns. 
overall_report_data = df1[['subtotal','shipping_price','total_price','item_cost','profit']].apply(currency_format_millions)
overall_report_data.rename(columns={
'subtotal' : 'Subtotal (millions)',
'shipping_price' : 'Shipping Price (millions)',
'total_price' : 'Total Price (millions)',
'item_cost' : 'Item Cost (millions)',
'profit' : 'Profit (millions)'
}, inplace=True)
overall_report_data
# Rename the columns to reflect the change in the money format. 




Subtotal (millions)  Shipping Price (millions)  Total Price (millions)  \
0                     0.12                       0.00                    0.13   
1                     0.00                       0.00                    0.00   
2                     0.00                       0.00                    0.00   
3                     0.00                       0.00                    0.00   
4                     0.00                       0.00                    0.00   
...                    ...                        ...                     ...   
54634                 0.00                       0.00                    0.00   
54635                 0.01                       0.00                    0.01   
54636                 0.03                       0.00                    0.03   
54637                 0.00                       0.00                    0.00   
54638                 0.00                       0.00                    0.00   

       Item Cost (millions)  Profit (millions)  
0                      0.08               0.05  
1                      0.00               0.00  
2                      0.00               0.00  
3                      0.00               0.00  
4                      0.00               0.00  
...                     ...                ...  
54634                  0.00               0.00  
54635                  0.01               0.00  
54636                  0.03               0.01  
54637                  0.00               0.00  
54638                  0.00               0.00  

[54639 rows x 5 columns]

In [440]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
overall_report_data.sort_values(by='Total Price (millions)', ascending=False).round(2)

Subtotal (millions)  Shipping Price (millions)  Total Price (millions)  \
21801               584.47                       0.00                  638.54   
15112               206.12                       0.00                  225.19   
23929               148.25                       0.00                  161.96   
25972                98.10                       0.00                  107.18   
11740                82.45                       0.00                   90.08   
...                    ...                        ...                     ...   
22529                 0.00                       0.00                    0.00   
37936                 0.00                       0.00                    0.00   
29972                 0.00                       0.00                    0.00   
21090                 0.00                       0.00                    0.00   
38399                 0.00                       0.00                    0.00   

       Item Cost (millions)  Profit (millions)  
21801                396.26             242.28  
15112                117.16             108.02  
23929                124.77              37.19  
25972                 71.38              35.80  
11740                 66.77              23.32  
...                     ...                ...  
22529                  0.00               0.00  
37936                  0.00               0.00  
29972                  0.00               0.00  
21090                  0.00               0.00  
38399                  0.00               0.00  

[54639 rows x 5 columns]

# Insight
##### Leadership was currious to get an insight as to based on the analyzed data which clients should be our target demographics within a certain profession
- Based on the current operating model and the our attractive price point, it seems to appear that customer with the following job titles Chief of Staff, Dietitian Osteopath, Pension scheme manager, and Production designer, theatre/television/film are our top customers. 
- Given this information, economically, it makese sense to invest, market and cross sell other products and services we offer to these target demographics

In [441]:
# round(df1.groupby('job')['qty'].sum().reset_index(), 2).sort_values
pd.set_option('display.float_format', lambda x: '%.2f' % x)
insight_data = pd.merge(df1, total_qty_spent, on='client_id')
insight_data = insight_data.groupby('job')['total_spent'].sum().reset_index()
insight_data

job    total_spent
0                                Chief of Staff  1673323498.77
1                                     Dietitian  1898783290.98
2                                     Osteopath  1845459870.41
3                        Pension scheme manager 15872445163.35
4  Production designer, theatre/television/film  1406442037.00